auth/api/iam/jar.go

```diff
@@ -0,0 +1,172 @@
+package iam
+
+import (
+	"context"
+	"crypto"
+	"errors"
+	"net/url"
+
+	"github.com/lestrrat-go/jwx/v2/jwt"
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/nuts-node/auth"
+	"github.com/nuts-foundation/nuts-node/auth/oauth"
+	cryptoNuts "github.com/nuts-foundation/nuts-node/crypto"
+	"github.com/nuts-foundation/nuts-node/vdr/resolver"
+)
+
+// requestObjectModifier is a function that modifies the Claims/params of an unsigned or signed (JWT) OAuth2 request
+type requestObjectModifier func(claims map[string]string)
+
+type requestObject struct {
+	Claims oauthParameters `json:"claims"`
+	Token  string          `json:"token,omitempty"`
+}
+
+func (ro requestObject) client() did.DID {
+	// set by JAR.Create, so always present
+	return did.MustParseDID(ro.Claims.get(oauth.ClientIDParam))
+}
+
+func (ro requestObject) nonce() string {
+	// set by JAR.Create, so always present
+	return ro.Claims.get(oauth.NonceParam)
+}
+
+func (ro requestObject) signed() bool {
+	return ro.Token != ""
+}
+
+var _ JAR = &jar{}
+
+type jar struct {
+	auth        auth.AuthenticationServices
+	jwtSigner   cryptoNuts.JWTSigner
+	keyResolver resolver.KeyResolver
+}
+
+type JAR interface {
+	// Create an unsigned request object.
+	// By default, it adds the following parameters:
+	// - client_id
+	// - jwt.Issuer
+	// - jwt.Audience (if server is not nil)
+	// - nonce
+	Create(client did.DID, server *did.DID, modifier requestObjectModifier) *requestObject
+	// Sign the requestObject, which is available on requestObject.Token.
+	// Returns an error if the requestObject already contains a signed JWT.
+	// TODO: check if signature type of client is supported by the AS/wallet.
+	Sign(ctx context.Context, ro *requestObject) error
```

Comment:
could return a signed requestobject

Reply:
signing now happens when the request object is fetched, so this now returns the token as sting. This change allows for consistent behavior in both get/post `request_uri_method`s. It also means the signed token is no longer part of the state / in storage. 

Signing now happens on the external call, and not the internal call. This probably means the audit on signing is in the wrong place again?